cmd/nephe-controller/main.go

```diff
@@ -47,6 +56,124 @@ func init() {
+	if changed := string(new.Data[key]) != string(old.Data[key]); changed {
+		return fmt.Errorf("secret data %s cannot be changed, old: %s , new: %s", key, new.Name, old.Name)
+	}
+	return nil
+}
+
+// fetchCPABySecret returns nil only when the secret is not used by cpa
+// otherwise, an appropriate error message and cpa object will be returned.
+func (v *secretValidator) fetchCPABySecret(newSecret *corev1.Secret, oldSecret *corev1.Secret, isDelete bool) (error, *crdv1alpha1.CloudProviderAccount) {
+	// TODO: Figure out how to get all CRs
+	cpaList := &crdv1alpha1.CloudProviderAccountList{}
+	options := &controllerclient.ListOptions{Limit: 1000}
+	err := v.client.List(context.TODO(), cpaList, options)
+	if err != nil {
+		return fmt.Errorf("failed to get cpa list, err:%v", err), nil
+	}
+
+	for _, cpa := range cpaList.Items {
+		setupLog.Info("Checking cpa", "name", cpa.Name, "namespace", cpa.Namespace)
+		if cpa.Spec.AWSConfig != nil {
+			if cpa.Spec.AWSConfig.SecretRef.Name == oldSecret.Name && cpa.Spec.AWSConfig.SecretRef.Namespace == oldSecret.Namespace {
+				if isDelete {
+					// Delete is not allowed.
+					setupLog.Info("Cannot delete secret", "name", oldSecret.Name, "cpa", oldSecret.Name)
+					return fmt.Errorf("cannot delete secret %s", oldSecret.Name), &cpa
+				}
+
+				if err = v.allowSecretUpdate(newSecret, oldSecret, cpa.Spec.AWSConfig.SecretRef.Key); err != nil {
+					return err, &cpa
+				}
+			}
+		}
+
+		if cpa.Spec.AzureConfig != nil {
+			if cpa.Spec.AzureConfig.SecretRef.Name == oldSecret.Name && cpa.Spec.AzureConfig.SecretRef.Namespace == oldSecret.Namespace {
+				if isDelete {
+					// Delete is not allowed.
+					setupLog.Info("Cannot delete secret", "name", oldSecret.Name, "cpa", oldSecret.Name)
+					return fmt.Errorf("cannot delete secret %s", oldSecret.Name), &cpa
+				}
+				if err = v.allowSecretUpdate(newSecret, oldSecret, cpa.Spec.AzureConfig.SecretRef.Key); err != nil {
+					return err, &cpa
+				}
+			}
+		}
+	}
+	return nil, nil
+}
+
+// Handle handles admission requests for secret.
+func (v *secretValidator) Handle(ctx context.Context, req admission.Request) admission.Response {
+	setupLog.Info("Received webhook", "req", req, "operation", req.Operation)
+	switch req.Operation {
+	case admissionv1.Create:
+		return admission.Allowed("")
+	case admissionv1.Update:
+		newSecret := &corev1.Secret{}
+		err := v.decoder.Decode(req, newSecret)
+		if err != nil {
+			setupLog.Error(err, "Error while decoding secretValidator", "secretValidator", req.Name)
+			return admission.Errored(http.StatusBadRequest, err)
+		}
+		oldSecret := &corev1.Secret{}
+		err = v.decoder.DecodeRaw(req.OldObject, oldSecret)
+		if err != nil {
+			setupLog.Error(err, "Error while decoding secretValidator", "secretValidator", req.Name)
+			return admission.Errored(http.StatusBadRequest, err)
+		}
+		err, cpa := v.fetchCPABySecret(newSecret, oldSecret, false)
+		if err != nil && strings.Contains(err.Error(), "failed to get cpa list") {
```

Comment:
what about other errors?

Reply:
This is to handle a case where we were unable to get CPA list (for whatever reason).
For other scenarios, we will always return a CPA CR or error will be nil.